Add Dockerfile, Script and Resources

sienkin opened the pull request:
Added Dockerfile to build container based on official `python:3.9-alpine` image
Added script and requirements file
Added README

Resolves: DVOP-880

sienkin committed:
Added Dockerfile and script resources to build container

sienkin committed:
Updated README

sienkin committed:
Fix README formatting

sienkin committed:
Fix README formatting

sienkin committed:
Added ability to change the output directory

sienkin committed:
Removed extraneous import

sienkin committed:
Added version file

sienkin committed:
Added missing table closing pipe

sienkin committed:
Removed unnecessary references to json2html module

sienkin committed:
Run python script through the  linter to standardise formatting

jamesgeddes commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,209 @@
+
+
+def generate_reporturl(ecr_scanfindings, region):
+    report_url = "https://{region}.console.aws.amazon.com/ecr/repositories/private/{registry}/{repository}/image/{digest}/scan-results/?region={region}".format(
```
This is longer than the [PEP8 permitted line length](https://www.python.org/dev/peps/pep-0008/#:~:text=spaces%20for%20indentation.-,Maximum%20Line%20Length,be%20limited%20to%2072%20characters.) of 79 characters.

sienkin replied:
Not really avoidable, unfortunately, as the string is constructing an AWS console URL.
The string could be broken up in to constituent parts but that would lead to lack of clarity.

jamesgeddes commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,209 @@
+        {
+            "fallback": heading,
+            "pretext": heading
+            + " for *<$ATC_EXTERNAL_URL/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|$BUILD_PIPELINE_NAME>*",
```
Long line length

sienkin replied:
This line contructs a Concourse URL and so can't really be split up without sacrificing readability

jamesgeddes commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,209 @@
+    return report_url
+
+
+def generate_slackjson(ecr_imagedata, vulnerabilities, report_url, report_dir):
```
Suggestion: Would `generate_slackjson` be more clear as `generate_slack_json`?

sienkin replied:
This is consistent with all other function names. In the interest of clarity, they've all been updated in 3faefe0

jamesgeddes commented:
LGTM, just a couple of nit-picky things

sienkin committed:
Updated function names for clarity

sienkin committed:
Reverted configurable output directory. Use of input_mapping in a Concourse pipeline will allow the configurable output directory

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+import time
+
+
+def logoutput(level, message):
```
I'd look to change this to `log_output` for consistency

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def get_severities(min_severity):
+    severities = ["CRITICAL", "HIGH", "MEDIUM", "LOW", "INFORMATIONAL"]
```
We should probably consider reworking these using enums but for the present I'd probably just make this list global

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+def get_severities(min_severity):
+    severities = ["CRITICAL", "HIGH", "MEDIUM", "LOW", "INFORMATIONAL"]
+
+    if min_severity == "ANY":
```
I'd be tempted to ditch the `ANY` as it's not really a log level and instead just give the log level a default of `None`. We could then replace this block with something like:

```
if level is None:
        return severities
    else:
        return severities[severities.index(level):]
```

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+
+
+if __name__ == "__main__":
+    # Ensure required environment vars are set
```
I'd pull this out into a separate function like [this](https://github.com/companieshouse/ois-tuxedo-qsp-transfer/blob/main/qsp_transfer.py#L18). It also has the added benefit of giving details of every variable that's missing in one hit rather than potentially having to hit each in turn

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+        ecr_client = session.client("ecr")
+
+    except session.exceptions.UnauthorizedException as err:
+        logoutput("Error", "Unable to authorise request. Invalid credentials")
```
If this is thrown due to authorisation then this message may be a little misleading. I.e. we may need to differentiate better between authentication and authorisation

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+    if not os.path.exists(report_dir):
+        os.mkdir(report_dir)
+
+    json_path = "./" + report_dir + "/report.json"
```
I'd template this for consistency. I.e. `f"./{report_dir}/report.json"`

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+        logoutput("Error", "Required AWS environment variables missing or not set")
+        sys.exit(1)
+
+    # Parse command line arguments/options
```
Again I'd pull this into a dedicated function

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,204 @@
+    ecr_imagedata = ecr_describe_scan_findings(ecr_client, args.imagerepo, args.imagetag)
+
+    # Wait and try again if the scan status is not COMPLETE
+    while ecr_imagedata["imageScanStatus"]["status"] != "COMPLETE":
```
I'd suggest a `MAX_TRY` or `MAX_WAIT` to guard against jobs running indefinitely

sienkin committed:
Updated log_output function name for consistency
Made list of severities global
Updated logic for get_severities function
Moved ENV var check to its own function and improved logic
Templated scan report output path
Moved parsing of command line args to a function
Implemented a timeout when waiting for scan results to complete (default 120 seconds)

sienkin committed:
Updated readme for new var

Jata2k commented on resources/ecr-scan-check.py:
```diff
@@ -0,0 +1,232 @@
+        ecr_client = session.client("ecr")
+
+    except session.exceptions.UnauthorizedException as err:
+        logoutput("Error", "Invalid SSO credentials")
```
I would change the error text here as the credentials aren't necessarily SSO. I.e. `SSO` -> `AWS`

sienkin committed:
Clarified error message
